clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java

```diff
@@ -31,32 +31,153 @@
+        List<String> topics = Arrays.asList("topic1", "topic2", "topic3", "topic4");
+        int recordSize = 3;
+        int partitionSize = 10;
+        int emptyPartitionIndex = 6;
+        int expectedTotalRecordSizeOfEachTopic = recordSize * (partitionSize - 1);
+
+        ConsumerRecords<Integer, String> consumerRecords = buildTopicTestRecords(recordSize, partitionSize, emptyPartitionIndex, topics);
+
+        for (String topic : topics) {
+            Iterable<ConsumerRecord<Integer, String>> records = consumerRecords.records(topic);
+            int recordCount = 0;
+            int partitionCount = 0;
+            int currentPartition = -1;
+
+            for (ConsumerRecord<Integer, String> record : records) {
+                validateEmptyPartition(record, emptyPartitionIndex);
+
+                // Check if we have moved to a new partition
+                if (currentPartition != record.partition()) {
+                    // Increment the partition count as we have encountered a new partition
+                    partitionCount++;
+                    // Update the current partition to the new partition
+                    currentPartition = record.partition();
+                }
+
+                validateRecordPayload(topic, record, currentPartition, recordCount, recordSize);
+                recordCount++;
+            }
+
+            // Including empty partition
+            assertEquals(partitionSize, partitionCount + 1);
+            assertEquals(expectedTotalRecordSizeOfEachTopic, recordCount);
+        }
+    }
+
+    private ConsumerRecords<Integer, String> buildTopicTestRecords(int recordSize,
+                                                                   int partitionSize,
+                                                                   int emptyPartitionIndex,
+                                                                   Collection<String> topics) {
+        Map<TopicPartition, List<ConsumerRecord<Integer, String>>> partitionToRecords = new LinkedHashMap<>();
+        for (String topic : topics) {
+            for (int i = 0; i < partitionSize; i++) {
+                List<ConsumerRecord<Integer, String>> records = new ArrayList<>(recordSize);
+                if (i != emptyPartitionIndex) {
+                    for (int j = 0; j < recordSize; j++) {
+                        records.add(
+                            new ConsumerRecord<>(topic, i, j, 0L, TimestampType.CREATE_TIME,
+                                0, 0, j, String.valueOf(j), new RecordHeaders(), Optional.empty())
+                        );
+                    }
+                }
+                partitionToRecords.put(new TopicPartition(topic, i), records);
+            }
+        }
+
+        return new ConsumerRecords<>(partitionToRecords);
+    }
+
+    private void validateEmptyPartition(ConsumerRecord<Integer, String> record, int emptyPartitionIndex) {
+        if (record.partition() == emptyPartitionIndex) {
```

Comment:
It can be `assertNotEquals(emptyPartitionIndex, record.partition(), "Partition " + record.partition() + " is not empty");`, right?

Reply:
Yes, in this way it could be more straightforward